File: app/src/test/java/com/mgaetan89/showsrage/presenter/ShowPresenterTest.java
package com.mgaetan89.showsrage.presenter;

import com.google.gson.Gson;
import com.mgaetan89.showsrage.model.Show;
import com.mgaetan89.showsrage.network.SickRageApi;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.Parameterized;

import java.util.Arrays;
import java.util.Collection;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(Parameterized.class)
public class ShowPresenterTest {
	@Parameterized.Parameter(1)
	public String bannerUrl;

	@Parameterized.Parameter(2)
	public int downloaded;

	@Parameterized.Parameter(3)
	public int episodesCount;

	@Parameterized.Parameter(4)
	public String network;

	@Parameterized.Parameter(5)
	public boolean paused;

	@Parameterized.Parameter(6)
	public String posterUrl;

	@Parameterized.Parameter(7)
	public String quality;

	@Parameterized.Parameter(0)
	public Show show;

	@Parameterized.Parameter(8)
	public String showName;

	@Parameterized.Parameter(9)
	public int snatched;

	private ShowPresenter presenter;

	@Before
	public void before() {
		this.presenter = new ShowPresenter(this.show);
	}

	@Test
	public void getBannerUrl() {
		assertThat(this.presenter.getBannerUrl()).isEqualTo(this.bannerUrl);
	}

	@Test
	public void getDownloaded() {
		assertThat(this.presenter.getDownloaded()).isEqualTo(this.downloaded);
	}

	@Test
	public void getEpisodesCount() {
		assertThat(this.presenter.getEpisodesCount()).isEqualTo(this.episodesCount);
	}

	@Test
	public void getNetwork() {
		assertThat(this.presenter.getNetwork()).isEqualTo(this.network);
	}

	@Test
	public void getPosterUrl() {
		assertThat(this.presenter.getPosterUrl()).isEqualTo(this.posterUrl);
	}

	@Test
	public void getQuality() {
		assertThat(this.presenter.getQuality()).isEqualTo(this.quality);
	}

	@Test
	public void getShowName() {
		assertThat(this.presenter.getShowName()).isEqualTo(this.showName);
	}

	@Test
	public void getSnatched() {
		assertThat(this.presenter.getSnatched()).isEqualTo(this.snatched);
	}

	@Test
	public void isPaused() {
		assertThat(this.presenter.isPaused()).isEqualTo(this.paused);
	}

	@Parameterized.Parameters
	public static Collection<Object[]> data() {
		Gson gson = SickRageApi.Companion.getGson();

		return Arrays.asList(new Object[][]{
				{null, "", 0, 0, "", false, "", "", "", 0},
				{gson.fromJson("{downloaded: 10, episodesCount: 20, network: \"ABC\", paused: 0, quality: \"HD\", show_name: \"Show 1\", snatched: 5, tvdbid: 123}", Show.class), "https://127.0.0.1:8083/api/apiKey/?cmd=show.getbanner&tvdbid=123", 10, 20, "ABC", false, "https://127.0.0.1:8083/api/apiKey/?cmd=show.getposter&tvdbid=123", "HD", "Show 1", 5},
				{gson.fromJson("{downloaded: 20, episodesCount: 30, network: \"CBS\", paused: 1, quality: \"HD1080p\", show_name: \"Show 2\", snatched: 10, tvdbid: 456}", Show.class), "https://127.0.0.1:8083/api/apiKey/?cmd=show.getbanner&tvdbid=456", 20, 30, "CBS", true, "https://127.0.0.1:8083/api/apiKey/?cmd=show.getposter&tvdbid=456", "HD1080p", "Show 2", 10},
		});
	}
}

File: app/src/test/kotlin/com/mgaetan89/showsrage/presenter/ShowPresenterTest.kt
package com.mgaetan89.showsrage.presenter

import com.google.gson.Gson
import com.mgaetan89.showsrage.model.Show
import com.mgaetan89.showsrage.network.SickRageApi
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.junit.runners.Parameterized
import org.mockito.Mockito.doReturn
import org.mockito.Mockito.spy

@RunWith(Parameterized::class)
class ShowPresenterTest(
		val show: Show?, val bannerUrl: String, val network: String, val paused: Boolean, val posterUrl: String,
		val quality: String, val showName: String
) {
	private lateinit var presenter: ShowPresenter

	@Before
	fun before() {
		this.presenter = spy(ShowPresenter(this.show))
		doReturn(this.show != null).`when`(this.presenter).isShowValid()
	}

	@Test
	fun getBannerUrl() {
		assertThat(this.presenter.getBannerUrl()).isEqualTo(this.bannerUrl)
	}

	@Test
	fun getNetwork() {
		assertThat(this.presenter.getNetwork()).isEqualTo(this.network)
	}

	@Test
	fun getPosterUrl() {
		assertThat(this.presenter.getPosterUrl()).isEqualTo(this.posterUrl)
	}

	@Test
	fun getQuality() {
		assertThat(this.presenter.getQuality()).isEqualTo(this.quality)
	}

	@Test
	fun getShowName() {
		assertThat(this.presenter.getShowName()).isEqualTo(this.showName)
	}

	@Test
	fun isPaused() {
		assertThat(this.presenter.isPaused()).isEqualTo(this.paused)
	}

	companion object {
		@JvmStatic
		@Parameterized.Parameters
		fun data(): Collection<Array<out Any?>> {
			val gson = SickRageApi.gson

			return listOf(
					arrayOf(null, "", "", false, "", "", ""),
					arrayOf(getShow(gson, "ABC", 0, "HD", "Show 1", 123), "https://127.0.0.1:8083/api/apiKey/?cmd=show.getbanner&tvdbid=123", "ABC", false, "https://127.0.0.1:8083/api/apiKey/?cmd=show.getposter&tvdbid=123", "HD", "Show 1"),
					arrayOf(getShow(gson, "CBS", 1, "HD1080p", "Show 2", 456), "https://127.0.0.1:8083/api/apiKey/?cmd=show.getbanner&tvdbid=456", "CBS", true, "https://127.0.0.1:8083/api/apiKey/?cmd=show.getposter&tvdbid=456", "HD1080p", "Show 2")
			)
		}

		private fun getShow(gson: Gson, network: String, paused: Int, quality: String, showName: String, tvDbId: Int): Show {
			return gson.fromJson("{network: \"$network\", paused: $paused, quality: \"$quality\", show_name: \"$showName\", tvdbid: $tvDbId}", Show::class.java)
		}
	}
}
